### DISCUSSION TITLE
Question, Pathfinder

### DISCUSSION THREAD

**Discussion by marcellangmaier** [NetTopologySuite/NetTopologySuite#520] [Q&A] [answered]
I am trying to get Pathfinder working for me. Similar to #381.

I am having a LineString network (representing a road) and two coordinates (P1, P2) somewhere near the LineStrings.
What I want to get is a LineString between those two points snapped to the road. (blue hand drawn line in the image below)

What I am doing so far is:
1. Using a GeometryNoder to create a graph
2. Add the graph to the Pathfinder
3. Snap the two points to the nearest locations on the LineString
4. Ask Pathfinder to connect the points
5. Getting **"Key not found in the graph"** exception

![seg_question](https://user-images.githubusercontent.com/26068556/117120079-2c4d6700-ad93-11eb-9301-bb29e108b80c.PNG)

I have created a simple project demonstrating my problem:
[TestPathfinder.zip](https://github.com/NetTopologySuite/NetTopologySuite/files/6426055/TestPathfinder.zip)

**Comment by FObermaier** [comment, marked as answer]
I assume you need to split the linestrings that are closest to your start- and endpoint, and then build the graph.

**Reply by marcellangmaier** [reply]
Thank you, you were right. I thought the GeometryNoder was doing this.
Instead of using the noder I simply added each vertex of each LineString in the geometry to the Pathfinder it started working.

**Comment by airbreather** [comment]
`PathFinder` only adds nodes for the **start** and **end** points of each `LineString`.

`DistanceOp` can return a coordinate **anywhere** in a `LineString`, even if it's noded.

You will need to use something else for snapping.  Suggestion, after you have `graph`:
```csharp
var kd = new KdTree<Coordinate>();
foreach (var ls in graph)
{
    kd.Insert(ls.GetCoordinateN(0));
    kd.Insert(ls.GetCoordinateN(ls.NumPoints - 1));
}
```

and then to get `p1Graph` / `p2Graph`:
```csharp
var p1Graph = kd.NearestNeighbor(p1_coord).Coordinate;
var p2Graph = kd.NearestNeighbor(p2_coord).Coordinate;
```

**Reply by marcellangmaier** [reply]
Thanks, the snapping was my second mistake which did not work as I imagined.
Fixed this with your advice.
